MAINT: slsqp callback accepts intermediate_result

andyfaff opened the pull request:
Toward gh-23570.

There were two holdouts of minimize callbacks not accepting an intermediate_result. SLSQP was one of them. This PR addresses that.

andyfaff committed:
MAINT: slsqp callback accepts intermediate_result

andyfaff committed:
MAINT: slsqp callback accepts intermediate_result

andyfaff committed:
STY

andyfaff committed:
STY

andyfaff committed:
STY

andyfaff committed:
MAINT: fix test

mdhaber commented:
I'll take a look today. Failures seem unrelated. If you recall whether there was a standard test we used to loop over all methods, it would be good to use that. (I'll look later.)

mattlevine22 commented:
Yes, can confirm these changes allowed SLSQP to use callback successfully both via the old `xk` and new `intermediate_result` interfaces.

andyfaff commented:
I added slsqp to `MINIMIZE_METHODS_NEW_CB`. However, `scipy.optimize.tests.test_optimize::TestOptimizeSimple::test_callback_stopiteration[2-slsqp]` fails (and 0+1). I think it's because of the way SLSQP deals with iteration counting and when the callback is called.

I modified the slsqp_py code to add print statements `print(f"{x=} in callback")` and `print(f"{x=} on exit")`
From running that test I see that the callback isn't actually being called when maxiter is set.

i.e. [`state_dict['iter'] > iter_prev`](https://github.com/scipy/scipy/blob/main/scipy/optimize/_slsqp_py.py#L527) is never True in the test, so the callback never stops the iteration, and  the asserts (assert res.fun == ref.fun, assert_equal(res.x, ref.x)) fail.

I know from the test that I added though that `StopIteration` does work.

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into slsqp

mdhaber committed:
TST: optimize.minimize: add 'slsqp' to common callback StopIteration test

mdhaber commented:
> I think the test is bogus for SLSQP because the termination criterion assumes that the halt point is always at the end of a major iteration.

I activated the test for SLSQP, giving it an exception for the part it's not passing. I'm not sure if the test is bogus for assuming that SLSQP would behave like all the other methods or if SLSQP's behavior is an unfortunate exception; it is what it is. In any case, adding SLSQP the test checks several valuable things that weren't tested here (iteration count, status code), and the excercise led to strengthening the check with a check of the error message and `success` attribute, which caught a bug in COBYQA (gh-23660).

If tests pass and if the updates look OK to someone, I'll go ahead and merge.

andyfaff commented on scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:
```diff
@@ -306,7 +306,7 @@ def _minimize_trustregion_constr(fun, x0, args, grad,
         * 1 : `gtol` termination condition is satisfied.
         * 2 : `xtol` termination condition is satisfied.
-        * 3 : `callback` function requested termination.
+        * 3 : `callback` raised `StopIteration`.
```
Won't this be an issue for downstream projects if we change the message text, code might be looking for specific text.

mdhaber replied:
I recall other discussions of changing error messages where the opinion has been that the text is not considered to be stable API and that users relying on them programmatically is bad practice. It seems that Python doesn't hesitate to change error messages, since there is a section "Improved error messages" several release notes (e.g. [3.10](https://docs.python.org/3/whatsnew/3.10.html#better-error-messages),  [3.12](https://docs.python.org/3/whatsnew/3.12.html#improved-error-messages), [3.13](https://docs.python.org/3/whatsnew/3.13.html#improved-error-messages)). NumPy has also documented changes in release note messages (e.g. [2.3](https://numpy.org/devdocs/release/2.3.0-notes.html#changes), [2.4](https://numpy.org/devdocs/release/2.4.0-notes.html#improved-error-message-for-assert-array-compare)). Instead, we provide `status` codes, and I've restrained myself from making those consistent.

andyfaff replied:
I'm fine with messages that aren't relied upon not to be strictly unchanging. Having said that, I think there's various items scattered around the scipy code base that do this kind of check.
The only thing that should definitely be consistent is a status code equalling zero for success

mdhaber replied:
Feel free to change that part back. It's just to make things be consistent but not worth holding up this PR.

> The only thing that should definitely be consistent is a status code equalling zero for success

Yeah, and enhancing the common test found that wasn't true for cobyqa (gh-23660).

andyfaff replied:
I think we should merge as is

mdhaber replied:
I would, but we have a prerelease test failure. Reverting the last commit to see if somehow that is the cause.

andyfaff reviewed (commented):
LGTM, apart from the question mark over the specific text.

mdhaber committed:
Revert "TST: optimize.minimize: add 'slsqp' to common callback StopIteration test"

This reverts commit d4fc6d0d46dd44af7810407a5d351feb2acbd116.

mdhaber committed:
Reapply "TST: optimize.minimize: add 'slsqp' to common callback StopIteration test"

This reverts commit 4f6ec9475c574af0d5028f29000a90c26fdddc21.

andyfaff commented:
It could've been a flake?

mdhaber commented:
Wow. It actually looks real.

andyfaff commented:
The failing test is 
```
FAILED scipy/optimize/tests/test_optimize.py::TestWorkers::test_smoke[CG-2] - ExceptionGroup: multiple unraisable exception warnings (6 sub-exceptions)
```
with messages along the lines of:
```
    | Traceback (most recent call last):
    |   File "/opt/hostedtoolcache/Python/3.12.11/x64/lib/python3.12/multiprocessing/resource_tracker.py", line 207, in _send
    |     self.ensure_running()
    |   File "/opt/hostedtoolcache/Python/3.12.11/x64/lib/python3.12/multiprocessing/resource_tracker.py", line 126, in ensure_running
    |     return self._reentrant_call_error()
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/opt/hostedtoolcache/Python/3.12.11/x64/lib/python3.12/multiprocessing/resource_tracker.py", line 70, in _reentrant_call_error
    |     raise ReentrantCallError(
    | multiprocessing.resource_tracker.ReentrantCallError: Reentrant call into the multiprocessing resource tracker
    | 
    | During handling of the above exception, another exception occurred:
    | 
    | Traceback (most recent call last):
    |   File "/opt/hostedtoolcache/Python/3.12.11/x64/lib/python3.12/multiprocessing/util.py", line 227, in __call__
    |     res = self._callback(*self._args, **self._kwargs)
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/opt/hostedtoolcache/Python/3.12.11/x64/lib/python3.12/multiprocessing/synchronize.py", line 88, in _cleanup
    |     unregister(name, "semaphore")
    |   File "/opt/hostedtoolcache/Python/3.12.11/x64/lib/python3.12/multiprocessing/resource_tracker.py", line 203, in unregister
    |     self._send('UNREGISTER', name, rtype)
    |   File "/opt/hostedtoolcache/Python/3.12.11/x64/lib/python3.12/multiprocessing/resource_tracker.py", line 213, in _send
    |     warnings.warn(
    | UserWarning: ResourceTracker called reentrantly for resource cleanup, which is unsupported. The semaphore object '/mp-w68sugdi' might leak.
    | 
    | The above exception was the direct cause of the following exception:
    | 
    | Traceback (most recent call last):
    |   File "/opt/hostedtoolcache/Python/3.12.11/x64/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 67, in collect_unraisable
    |     warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
    | pytest.PytestUnraisableExceptionWarning: Exception ignored in: <Finalize object, dead>
    | Enable tracemalloc to get traceback where the object was allocated.
    | See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.
```

The test is using `multiprocessing`, so the presence of those ResourceTracker warnings seems related. I can't understand why the test is not failing for this PR and not others. There's nothing in this PR that is related to the "failing" test.

mdhaber committed:
TST: optimize.minimize: comment out failing test case

andyfaff commented:
https://github.com/nedbat/coveragepy/issues/1306 might be related.

andyfaff commented:
So I wonder if the test would pass if we didn't install pytest-cov

mdhaber committed:
TST: optimize.minimize: comment out different failing test cas

mdhaber commented:
Interesting. The test that was failing is passing after commenting out `trust-constr`. If the rest pass, I'll merge this and open a PR trying to bring back `trust-constr`. If that passes, great; if not, I'll close the PR and open an issue.

mdhaber commented:
Now it's failing in a different job : / Let's see if it starts appearing in other PRs.

andyfaff commented:
I wonder if it's due to any bumps in upstream packages.  pytest/pytest-cov were updated a couple of weeks ago.

I have some ideas about things to investigate further, but at the moment I can't reproduce it locally.

mdhaber committed:
Apply suggestions from code review

[skip ci]
